### DISCUSSION TITLE
rec模型训练时，将训练图进行高斯模糊、运动模糊、低分辨率缩放 以及 对抗性训练等增强 的完整python代码

### DISCUSSION THREAD

**Discussion by nissansz** [PaddlePaddle/PaddleOCR#14644] [Q&A]
rec模型训练时，将训练图进行高斯模糊、运动模糊、低分辨率缩放 以及 对抗性训练等增强 的完整python代码

**Comment by GreatV** [comment]
以下是完整的 Python 代码，可以在训练识别（rec）模型时对训练图像进行 **高斯模糊、运动模糊、低分辨率缩放** 以及 **对抗性训练** 进行数据增强：

```python
import cv2
import numpy as np
import random
import torch
import torch.nn.functional as F

# 高斯模糊
def apply_gaussian_blur(image):
    kernel_size = random.choice([3, 5, 7])  # 随机选择核大小
    return cv2.GaussianBlur(image, (kernel_size, kernel_size), 0)

# 运动模糊
def apply_motion_blur(image):
    size = random.choice([3, 5, 7])  # 运动模糊核大小
    kernel = np.zeros((size, size))
    xs, ys = np.random.choice(size, 2)  # 随机选择模糊方向
    kernel[xs, :] = 1
    kernel /= size
    return cv2.filter2D(image, -1, kernel)

# 低分辨率缩放
def apply_low_resolution(image):
    scale_factor = random.uniform(0.5, 0.8)  # 随机缩小比例
    h, w = image.shape[:2]
    new_w, new_h = int(w * scale_factor), int(h * scale_factor)
    image = cv2.resize(image, (new_w, new_h), interpolation=cv2.INTER_LINEAR)
    return cv2.resize(image, (w, h), interpolation=cv2.INTER_NEAREST)

# 对抗性训练（FGSM攻击）
def fgsm_attack(image_tensor, epsilon, gradient):
    """ FGSM 对抗样本攻击 """
    sign_grad = gradient.sign()
    perturbed_image = image_tensor + epsilon * sign_grad
    return torch.clamp(perturbed_image, 0, 1)  # 限制像素值范围

# 训练数据增强流程
def augment_image(image):
    """ 对图像应用一系列增强操作 """
    if random.random() < 0.3:
        image = apply_gaussian_blur(image)
    if random.random() < 0.3:
        image = apply_motion_blur(image)
    if random.random() < 0.3:
        image = apply_low_resolution(image)
    return image

# 示例：对抗性训练示例
def adversarial_training(model, image_tensor, label, epsilon=0.03):
    """ 生成对抗性样本并训练模型 """
    image_tensor.requires_grad = True
    output = model(image_tensor)
    loss = F.cross_entropy(output, label)
    loss.backward()
    
    adv_image = fgsm_attack(image_tensor, epsilon, image_tensor.grad)
    return adv_image

# 示例代码：应用增强
if __name__ == "__main__":
    image = cv2.imread("example.jpg")  # 读取训练图像
    image = augment_image(image)  # 应用数据增强
    cv2.imwrite("augmented_example.jpg", image)  # 保存增强后的图像

    # 假设有一个训练模型（仅示意）
    class DummyModel(torch.nn.Module):
        def forward(self, x):
            return torch.randn((1, 10), requires_grad=True)

    model = DummyModel()
    image_tensor = torch.randn((1, 3, 32, 32), requires_grad=True)  # 伪造的输入图像
    label = torch.tensor([1])  # 假设的标签
    adv_image = adversarial_training(model, image_tensor, label)
```

### 代码说明：
- **高斯模糊**：使用 `cv2.GaussianBlur()` 进行模糊。
- **运动模糊**：使用卷积核 `cv2.filter2D()` 进行模拟。
- **低分辨率缩放**：先缩小再放大，模拟低分辨率的情况。
- **对抗性训练**：使用 FGSM（Fast Gradient Sign Method）攻击生成对抗样本。

### 适用场景：
- **数据增强**：在 OCR 训练过程中提升模型的鲁棒性。
- **对抗性训练**：提升模型对攻击的抗干扰能力。

你可以将这些方法集成到 PaddleOCR 训练流程中，在 `dataset.py` 或者 `data_augment.py` 中调用这些数据增强方法，以提高模型的泛化能力。

---
*Response generated by [feifei-bot](https://github.com/GreatV/feifei) | chatgpt-4o-latest*
